tools/clang/lib/SPIRV/SPIRVEmitter.cpp

```diff
@@ -6798,6 +6823,34 @@ uint32_t SPIRVEmitter::translateAPInt(const llvm::APInt &intValue,
   return 0;
 }
 
+bool SPIRVEmitter::canBeRepresentedIn32Bits(const Expr *expr) {
+  if (const auto *intLiteral = dyn_cast<IntegerLiteral>(expr)) {
+    const bool isSigned = expr->getType()->isSignedIntegerType();
+    const llvm::APInt &value = intLiteral->getValue();
+    return (isSigned && value.isSignedIntN(32)) ||
+           (!isSigned && value.isIntN(32));
+  }
+
+  if (const auto *floatLiteral = dyn_cast<FloatingLiteral>(expr)) {
+    llvm::APFloat value = floatLiteral->getValue();
```

Comment:
Nit: Can move this assignment after the check for half and single float semantics so we won't need to copy the value for these two cases.